### DISCUSSION TITLE
Not on techempower benchmarks?

### DISCUSSION THREAD

**Discussion by ibsusu** [uNetworking/uWebSockets#1215] [General]
So I know that this project is specifically for web sockets but the speed is also very much there for normal http.  Why isn't it on techempower benchmarks?

**Comment by ghost** [comment]
Because tech empower are retards.

**Reply by ibsusu** [reply]
Would you care to elaborate?  To what degree are they retarded?  Is it infectious?  Are the people who submit their frameworks more or less retarded? （。々°）

**Reply by ghost** [reply]
I have explained this about 17 times by now but in essence, science doesn't work the way they think it does.

They measure multiple variables at the same time, attributing insignificant leads to the http server even though all of their tests really measure the database. All top 70 entries are clustered around PostgreSql but they are too illiterate and scientifically uneducated to see this.

Also they don't understand the concept of significance vs insignificance - if one test puts one multi-variable entry 0.000000001% above the other they write a blog post about the "winner' instead of concluding a draw and looking at improving their accuracy.

They further on don't understand the concept of CPU time.

Essentially they are uneducated, illiterate morons who have no idea about what they are doing

**Reply by ghost** [reply]
I reported an issue proving very obviously that all of their tests were nonsensical in round 12 or something and they agreed (their personnel agreed!) Yet they brushed it off as nothing. Who the fick just ignores fundamental flaws without giving a fick? Well someone who doesn't have passion for quality and accuracy. They just want to crown whomever as ultimate winner even though they essentially magnify noise.

They are so good damn ignorant that they would have magnified the background radiation and used that to find a winner and they wouldn't have noticed

**Reply by ghost** [reply]
In summation i wouldn't even trust tech empower to store my winter tires. They are the definition of illiteracy

**Reply by ibsusu** [reply]
OK so I get what your saying but after reading what you've written inside of the benchmarks directory what they should be doing is stress testing each of the servers with enough input to ensure 100% CPU time which means no waiting on a DB or the NIC.  My assumption would be that using a modified [load_test.c](https://github.com/uNetworking/uWebSockets/blob/master/benchmarks/load_test.c) that you have on more cores on the same machine than the server that's being tested would be the best way to truly find out which server was the best in terms of performance.

Yea or nay?

**Reply by ibsusu** [reply]
Obviously any confounding factors such as the db need to be separated from the metrics that we care about and be measured on their own.

**Reply by ibsusu** [reply]
I find this kinda funny.  What you've essentially done is get uWebSockets to the point where it has as little overhead as possible versus the kernel networking stack itself but the more I think about it the more I realize that most people's clumsy attempts to utilize uWebSockets will go to waste.  In my project I have to talk to the db, I have to write to disk and uws stays out of my way.  It's always my crap that slows it down.

**Reply by ghost** [reply]
The nic is to the web server what the hdd is to the database; you can include them in your test but if those become the bottleneck you're not really measuring the database or the web server you're measuring hardware.

Software runs on CPU-time so if you want to isolate and measure the CPU-time overhead of the web server or database (such as algorithmic complexities, memory allocations, etc) you have to put the CPU-time as the bottleneck.

Sometimes you can do that easily, such as on the raspberry pi where CPU-time is naturally limited.

Tech empower puts their measurements on Intel gold server CPUs where no software will ever bottleneck on CPU-time so they pretty much the up measuring their Nic (which is a constant) so the only difference between the top 30 entries become random noise.

With databases you also have scheduling and packing to take into account but we're not talking about databases here.

**Reply by ghost** [reply]
Yes performance is always the weakest link of a chain. USING something fast doesn't MAKE everything fast it only unlocks the POTENTIAL to MAKE everything fast.

**Reply by jobs-git** [reply]
Well, you are free to perform the benchmark yourself to validate which framework you will use. But what they did are still useful, as it gives developers an idea, but you will still have to test in on your production.

**Reply by uNetworkingAB** [reply]
TechEmpower is useful as a **negative**. That is, if something scores very low in TechEmpower it is a strong hint of poor performance. But the other way around is not true.

Top-scorers in TechEmpower are almost always:

* Entirely broken in every testable sense, such as failing the majority of h1spec tests (see my github for repo)
* Often times fully hacked together with the most minimal non-compliant parser possible
* Fast for pipelined requests but very slow for normal requests
* Riddled with security flaws, undefined behavior, etc

So TechEmpower can tell you what not to pick, but it can't tell you what to pick

**Comment by jobs-git** [comment]
> So I know that this project is specifically for web sockets but the speed is also very much there for normal http. Why isn't it on techempower benchmarks?

for it to be included, somebody should contribute and maintain the uwebsocket code

**Reply by uNetworkingAB** [reply]
It already is included. It scores top 3-ish. But like stated in this thread, TechEmpower is a test of shit quality and lacks the most basic credibility:

https://x.com/uNetworkingAB/status/1811425564764610926
